pkg/clustermesh/remote_cluster.go

```diff
@@ -0,0 +1,197 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clustermesh
+
+import (
+	"fmt"
+	"path"
+	"time"
+
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/ipcache"
+	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/kvstore/allocator"
+	"github.com/cilium/cilium/pkg/kvstore/store"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/node"
+
+	"github.com/sirupsen/logrus"
+)
+
+// remoteCluster represents another cluster other than the cluster the agent is
+// running in
+type remoteCluster struct {
+	// name is the name of the cluster
+	name string
+
+	// configPath is the path to the etcd configuration to be used to
+	// connect to the etcd cluster of the remote cluster
+	configPath string
+
+	// changed receives an event when the remote cluster configuration has
+	// changed and is closed when the configuration file was removed
+	changed chan bool
+
+	// mesh is the cluster mesh this remote cluster belogs to
+	mesh *ClusterMesh
+
+	controllers *controller.Manager
+
+	// remoteConnectionControllerName is the name of the backing controller
+	// that maintains the remote connection
+	remoteConnectionControllerName string
+
+	// mutex protects the following variables
+	// - store
+	// - remoteNodes
+	// - ipCacheWatcher
+	// - remoteIdentityCache
+	mutex lock.RWMutex
+
+	// store is the shared store representing all nodes in the remote cluster
+	remoteNodes *store.SharedStore
+
+	// ipCacheWatcher is the watcher that notifies about IP<->identity
+	// changes in the remote cluster
+	ipCacheWatcher *ipcache.IPIdentityWatcher
+
+	// remoteIdentityCache is a locally cached copy of the identity
+	// allocations in the remote cluster
+	remoteIdentityCache *allocator.RemoteCache
+
+	// backend is the kvstore backend being used
+	backend kvstore.BackendOperations
+}
+
+var (
+	// skipKvstoreConnection skips the etcd connection, used for testing
+	skipKvstoreConnection bool
+)
+
+func (rc *remoteCluster) getLogger() *logrus.Entry {
+	rc.mutex.RLock()
+	defer rc.mutex.RUnlock()
+
+	status, err := rc.backend.Status()
+
+	return log.WithFields(logrus.Fields{
+		"name":          rc.name,
+		"config":        rc.configPath,
+		"kvstoreStatus": status,
+		"kvstoreErr":    err,
+	})
+}
+
+func (rc *remoteCluster) restartRemoteConnection() {
+	rc.controllers.UpdateController(rc.remoteConnectionControllerName,
+		controller.ControllerParams{
+			DoFunc: func() error {
+				backend, err := kvstore.NewClient("etcd", map[string]string{
```

Comment:
Export the constants from `pkg/kvstore/etcd.go` and use them here: `etcdName`, `addrOption`, `cfgOption`.